src/emu/dinvram.h

```diff
@@ -38,15 +46,43 @@ class device_nvram_interface : public device_interface
 	void nvram_load(emu_file &file) { nvram_read(file); }
 	void nvram_save(emu_file &file) { nvram_write(file); }
 
+	bool has_battery_backup() const { return device_has_battery_backup(); }
+
+	memory_array &memarray() { return m_memarray; }
+
 protected:
 	// derived class overrides
 	virtual void nvram_default() = 0;
-	virtual void nvram_read(emu_file &file) = 0;
-	virtual void nvram_write(emu_file &file) = 0;
+	virtual void nvram_read(emu_file &file);
+	virtual void nvram_write(emu_file &file);
+	virtual bool device_has_battery_backup() const { return true; }
+
+private:
+	memory_array m_memarray;
 };
 
 // iterator
 typedef device_interface_iterator<device_nvram_interface> nvram_interface_iterator;
 
+// ======================> device_optional_nvram_interface
+
+// interface subclass for memories that may or may not be battery-backed
+class device_optional_nvram_interface : public device_nvram_interface
```

Comment:
This name is very confusing.  If you're going to keep this separate from the base `device_nvram_interface` it needs a better name.